lib/recursive-methods.rb

```diff
@@ -1,49 +1,88 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)?
+# Space complexity: O(n)?
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return n if n == 1
+  return 1 if n == 0
+  raise ArgumentError if n < 0
+
+  n = n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) ?
+# Space complexity: O(n) ?
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  return s[-1] + reverse(s[0...-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) ?
+# Space complexity: O(n) ?
 def reverse_inplace(s)
```

Comment:
Because you are creating a new array with each recursive call, this is O(n^2)  in space complexity.

It's also definitely **not** in place because you're creating new arrays.

Reply:
I'm confused. Where are I creating a new array? You can find characters in a string my their index. That is all I am doing. If the string is a = "hello" a[1] == "e" and a[0..-1] == "hello".  And you can add strings b = "dolly" . So a + b = "hellodolly".